warehouse/oidc/views.py

```diff
@@ -0,0 +1,98 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from pyramid.view import view_config
+
+from warehouse.admin.flags import AdminFlagValue
+from warehouse.events.tags import EventTag
+from warehouse.macaroons import caveats
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.oidc.interfaces import IOIDCProviderService
+
+
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+)
+def mint_token_from_oidc(request):
+    def _invalid(errors):
+        request.response.status = 422
+        return {"message": "Token request failed", "errors": errors}
+
+    oidc_enabled = request.registry.settings[
+        "warehouse.oidc.enabled"
+    ] and not request.flags.enabled(AdminFlagValue.DISALLOW_OIDC)
+    if not oidc_enabled:
+        return _invalid(
+            errors=[
+                {
+                    "code": "not-enabled",
+                    "description": "OIDC functionality not enabled",
+                }
+            ]
+        )
+
+    try:
+        body = request.json_body
+    except ValueError:
+        return _invalid(
+            errors=[{"code": "invalid-json", "description": "missing JSON body"}]
+        )
+
+    unverified_jwt = body.get("token")
```

Comment:
We actually _do_ catch JSON decoding errors, since the `json_body` property raises if the request body isn't valid JSON.

```python
    try:
        body = request.json_body
    except ValueError:
        return _invalid(
            errors=[{"code": "invalid-json", "description": "missing JSON body"}]
        )
```

(Ref: https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/webob.html#request-json-body)

In this case it's a type error, since payload _is_ valid JSON but the top-level JSON type is a string instead of a map. So I'll add another check here to make sure we're getting the type we actually expect out of `json_body` 🙂

Reply:
0f03cab7

Reply:
Ah, good catch. I can confirm that this fixes the issue.